pyop2/mpi.py

```diff
@@ -206,44 +407,62 @@ def free_comm(comm, remove=True):
             else:
                 ocomm.Delete_attr(innercomm_keyval)
             del icomm
-        if remove:
-            # Only do this if not called from free_comms.
+        try:
             dupped_comms.remove(comm)
+        except ValueError:
+            debug(f"{comm.name} is not in list of known comms, probably already freed")
+            debug(f"Known comms are {[d.name for d in dupped_comms if d != MPI.COMM_NULL]}")
         compilation_comm = get_compilation_comm(comm)
-        if compilation_comm is not None:
-            compilation_comm.Free()
+        if compilation_comm == MPI.COMM_NULL:
+            comm.Delete_attr(compilationcomm_keyval)
+        elif compilation_comm is not None:
+            free_comm(compilation_comm)
+            comm.Delete_attr(compilationcomm_keyval)
         comm.Free()
+    else:
+        warning('Attempt to free MPI_COMM_NULL')
 
 
 @atexit.register
-def free_comms():
+def _free_comms():
     """Free all outstanding communicators."""
+    global PYOP2_FINALIZED
+    PYOP2_FINALIZED = True
+    if logger.level > DEBUG:
+        debug = lambda string: None
+    else:
+        debug = lambda string: print(string)
+    debug("PyOP2 Finalizing")
+    # Collect garbage as it may hold on to communicator references
+    debug("Calling gc.collect()")
+    import gc
+    gc.collect()
+    debug(pyop2_comm_status())
+    debug(f"Freeing comms in list (length {len(dupped_comms)})")
     while dupped_comms:
-        c = dupped_comms.pop()
-        refcount = c.Get_attr(refcount_keyval)
-        for _ in range(refcount[0]):
-            free_comm(c, remove=False)
+        c = dupped_comms[-1]
+        if is_pyop2_comm(c):
```

Comment:
Do we ever free non-PyOP2 communicators?

Reply:
Let's find out...

Reply:
Apparently not, at least not in the PyOP2 test suite